repository/fcm_token_repository.py

```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
+from domain.entity.fcm_tokens import FCMToken
+from domain import session_scope, User
+import logging
+
+
+class FCMTokenRepository:
+    def __init__(self):
+        pass
+
+    def check_user_exists(self, user_id):
+
+        # Check if a user with the given user_id exists
+        with session_scope() as session:
+            user = session.query(User).filter_by(id=user_id).first()
+            return user is not None
+
+    def register_token(self, user_id, fcm_token, device_type):
+
+        with session_scope() as session:
+            try:
+                # 1. Check if there is user given userId
+                existing_token = session.query(FCMToken).filter_by(user_id=user_id, fcm_token=fcm_token).first()
+
+                if existing_token:
+                    existing_token.updated_at = datetime.now()
+                    session.commit()
+                    logging.info(f" A New token is registered for user {user_id}")
+                else:
+                    new_token = FCMToken(
+                        user_id=user_id,
+                        fcm_token=fcm_token,
+                        device_type=device_type,
+                        created_at=datetime.now(),
+                        updated_at=datetime.now()
+                    )
+                    session.add(new_token)
+                    session.commit()
+                    logging.info(f" A New token is registered for user {user_id}")
+
+            except Exception as e:
+                logging.error(f"Error registering FCM token for user {user_id}: {e}")
+                session.rollback()
+
+    def unregister_token(self, user_id, fcm_token):
+
+        with session_scope() as session:
+            try:
+                existing_token = session.query(FCMToken).filter_by(user_id=user_id, fcm_token=fcm_token).first()
+
+                if existing_token:
+                    session.delete(existing_token)
+                    session.commit()
+                    logging.info(f" Unregistered the token for user {user_id}")
+                    return True
+                else:
+                    logging.info(f" Invalid token for user {user_id}")
+                    return False
```

Comment:
I would prefer if you threw a custom exception when there is an invalid token rather than returning `False`. To be clear, your code is totally functional, but an exception is supposed to represent an error in the program flow, including when invalid data is provided. I think you had this instinct too, since in the API layer you translate a `False` into a 400 error.

Making this change will require you to be more specific in what exceptions you catch on line 60, change it to [`SQLAlchemyError`](https://docs.sqlalchemy.org/en/20/core/exceptions.html#sqlalchemy.exc.SQLAlchemyError).

Reply:
To be clear, in the API layer you should just change the logic for when you return a 400 to just be a catch of your custom exception, maintaining the existing, general, exception catch that returns a 500.

Reply:
I am not confident with my new implementation according to this.